### PR TITLE
add spec for existing active record object

### PR DESCRIPTION
**Pull request by palkan**
Hi! 
Just added a test case for handling newly closed bug (8561b7fc11374a9d396318921618894eef3f8d76).

### CODE REVIEW

**Comment by RobertGauld** [comment]
Looks good, I couldn't work out how to test the fix #LearningFromEachOther. 
I'll accept when I get home unless someone beats me to it.

**Comment by einzige** [comment]
@palkan Isn't this https://github.com/palkan/date_time_attribute/blob/fix-active-record-bug/spec/active_record_spec.rb#L8 and this https://github.com/palkan/date_time_attribute/blob/fix-active-record-bug/spec/active_record_spec.rb#L37 perform the same thing you contribute?

Not sure if this test case covers https://github.com/einzige/date_time_attribute/commit/8561b7fc11374a9d396318921618894eef3f8d76

It was originally covered here https://github.com/einzige/date_time_attribute/commit/923f7c2418f4e9f606197151014031bac0a3b658

File: spec/active_record_spec.rb [hunk 1]
```diff
@@ -100,4 +100,25 @@
       end
     end
   end
+
+
+  context "existed record" do
+
```
Comment (einzige) [inline spec/active_record_spec.rb]: follow formatting, remove empty lines

File: spec/active_record_spec.rb [hunk 2]
```diff
@@ -100,4 +100,25 @@
...
+    subject(:target) { Model.first}
```
Comment (einzige) [inline spec/active_record_spec.rb]: formatting, `{ Model.first`[space]`}`
Calling `Model.first` in tests is flaky. There is no reason for doing this. The record not necessarily should be the first..

the correct code is:

``` ruby
subject(:target) { Model.create created_at: 'whatever you set' }
```

File: spec/active_record_spec.rb [hunk 3]
```diff
@@ -100,4 +100,25 @@
...
+
+    describe 'values' do
+      subject(:date_time) { target.created_at }
+
+      context 'datetime set' do
+        it { should_not be_nil }
```
Comment (einzige) [inline spec/active_record_spec.rb:127]: https://github.com/palkan/date_time_attribute/blob/fix-active-record-bug/spec/active_record_spec.rb#L39 same here

File: spec/active_record_spec.rb [hunk 4]
```diff
@@ -100,4 +100,25 @@
...
+      end
+
+      context 'datetime correct' do
+        its(:year) { should == 2014 }
```
Comment (einzige) [inline spec/active_record_spec.rb]: by doing this you basically check ActiveRecord's implementation of attribute setter/getters. It is not related to this gem at all. You assigned the datetime and you test if it is assigned properly.

File: spec/spec_helper.rb [hunk 5]
```diff
@@ -14,6 +14,9 @@ class Model < ActiveRecord::Base
     include DateTimeAttribute
     date_time_attribute :created_at
   end
+
+  Model.create! created_at: '2014-01-01 12:00:00 +0000'
```
Comment (einzige) [inline spec/spec_helper.rb]: Flaky. You pollute test suite with shared data. You should move test-specific code into test implementation and never to spec_helper. It is hard to realise where this record is gonna be used.

File: spec/active_record_spec.rb (hunk 2 above)
Comment (RobertGauld) [inline spec/active_record_spec.rb]: So being home gives me a larger screen and the chance to better look at the code.

Although that will simply create a new model, not test a model loaded from the database  (the original bug) (the bit where I got a bit stuck with the test, thank palkan for the push to get unstuck). I suppose a combination of what was added to the helper and what you say here would be the way forward.

``` ruby
Model.create! id: 1, created_at: '2014-01-01 12:00:00 +0000'
subject(:target) { Model.find(1) }
```

Or even 

``` ruby
Model.create! created_at: '2014-01-01 12:00:00 +0000' # Make sure at least one model exists
subject(:target) { Model.first }
```

and simply check that what the accessors return isn't nil (the original problem) but move creating the model closer to the actual test.

File: spec/active_record_spec.rb (hunk 4 above)
Comment (RobertGauld) [inline spec/active_record_spec.rb]: Thus not testing what is retrieved from the database.

**Comment by RobertGauld** [comment]
@palkan thanks for the contribution (and giving me the chance to learn by being unstuck) but given the above comments (and as it's einzige's project) I think that at least the following need to be resolved before I would feel comfortable accepting it on einzige's behalf.
- Moving creation of the model from the helper and closer to (or preferably into) the test
- Not setting the value before checking it's (and the _date, _time methods) not nil
- Removing of empty lines to copy the format of the rest of the file

**Comment by RobertGauld** [comment]
@einzige "Isn't this https://github.com/palkan/date_time_attribute/blob/fix-active-record-bug/spec/active_record_spec.rb#L8 and this https://github.com/palkan/date_time_attribute/blob/fix-active-record-bug/spec/active_record_spec.rb#L37 perform the same thing you contribute?"

No those test a model which has been created and saved to the database not one which has been read from the database. In the bug the values of field, field_date, field_time were nil after the record was read from the database.

Also it's a test for https://github.com/einzige/date_time_attribute/commit/8561b7fc11374a9d396318921618894eef3f8d76

File: spec/active_record_spec.rb (hunk 2 above)
Comment (einzige) [inline spec/active_record_spec.rb]: gotcha, thanks! @palkan So all we have to do is:

``` ruby
subject(:target) { Model.create!(created_at: '2014...').reload }
```

and let's change the context name to make it more clear:

``` ruby
context 'loaded from db' 
```

or to something more meaningful.

**Comment by einzige** [comment]
`create!` should reload record from db.. since it needs new data after INSERT is performed.
Anyway, I will check if specs were failing with new gemfile and no changes in gem code.

btw, @robertgauld did this https://github.com/einzige/date_time_attribute/commit/8561b7fc11374a9d396318921618894eef3f8d76 change specs results? In this PR we implemented specs without changing the code and they are still passing.

**Comment by palkan** [comment]
@einzige  the following approach doesn't do the trick

``` ruby
subject(:target) { Model.create!(created_at: '2014...').reload }
```

because  [`reload`](http://apidock.com/rails/ActiveRecord/Persistence/reload) returns `self` with "in-place modifications" (which are absent in the case). So we have to make some sort of `find` to _really_ read the record from db.

**Comment by RobertGauld** [comment]
No, https://github.com/einzige/date_time_attribute/commit/8561b7fc11374a9d396318921618894eef3f8d76 did not change spec results.

Also I can confirm that when running the buggy version

``` ruby
Model.create!(created_at: '2014...').reload.created_at  # doesn't return nil
Model.create!(created_at: '2014...').reload.created_at_date  # doesn't return nil
Model.create!(created_at: '2014...').reload.created_at_time  # doesn't return nil

Model.first.created_at  # returns nil
Model.first.created_at_date  # returns nil
Model.first.created_at_time  # returns nil
```

**Comment by einzige** [comment]
@palkan @robertgauld let's make specs fail first to make sure they are really working properly.

**Comment by einzige** [comment]
ok, guys, I made them fail: https://travis-ci.org/einzige/date_time_attribute/builds/20195582
Merging now, thanks a lot.

Going forward, let's make sure specs are failing before we change the code.

File: spec/active_record_spec.rb [hunk 6]
```diff
@@ -1,6 +1,22 @@
 require 'spec_helper'
+require 'active_record'
 
-describe DateTimeAttribute, ActiveRecord::Base, use_active_record: true do
+ActiveRecord::Base.establish_connection(adapter: 'sqlite3', database: ':memory:')
```
Comment (einzige) [inline spec/active_record_spec.rb:4]: There was no reason of moving this from spec_helper. Moving back.

**Pull request merged**